pkg/agent/monitoraddnodes.go

```diff
@@ -0,0 +1,239 @@
+	nodeIPAddress string
+	cluster       *Cluster
+	status        addNodeStatusHistory
+}
+
+func newAddNodeMonitor(nodeIP string, cluster *Cluster) *addNodeMonitor {
+	mon := addNodeMonitor{
+		nodeIPAddress: nodeIP,
+		cluster:       cluster,
+		status: addNodeStatusHistory{
+			RestAPISeen:            false,
+			KubeletIsRunningOnNode: false,
+			FirstCSRSeen:           false,
+			SecondCSRSeen:          false,
+			NodeJoinedCluster:      false,
+			NodeIsReady:            false,
+		},
+	}
+	return &mon
+}
+
+func (mon *addNodeMonitor) logStatus(status string) {
+	logrus.Infof("Node %s: %s", mon.nodeIPAddress, status)
+}
+
+// MonitorAddNodes waits for the a node to be added to the cluster
+// and reports its status until it becomes Ready.
+func MonitorAddNodes(cluster *Cluster, nodeIPAddress string) error {
+	parsedIPAddress := net.ParseIP(nodeIPAddress)
+	if parsedIPAddress == nil {
+		return fmt.Errorf("%s is not valid IP Address", nodeIPAddress)
+	}
+
+	timeout := 90 * time.Minute
+	waitContext, cancel := context.WithTimeout(cluster.Ctx, timeout)
+	defer cancel()
+
+	mon := newAddNodeMonitor(nodeIPAddress, cluster)
+
+	wait.Until(func() {
+		if !mon.status.RestAPISeen &&
+			mon.cluster.API.Rest.IsRestAPILive() {
+			mon.status.RestAPISeen = true
+			mon.logStatus("Assisted Service API is available")
+		}
+
+		if !mon.status.KubeletIsRunningOnNode &&
+			mon.isKubeletRunningOnNode() {
+			mon.status.KubeletIsRunningOnNode = true
+			mon.logStatus("Kubelet is running")
+		}
+
+		if mon.status.KubeletIsRunningOnNode &&
+			!mon.status.FirstCSRSeen &&
+			mon.clusterHasFirstCSRPending() {
+			mon.status.FirstCSRSeen = true
+			mon.logStatus("First CSR Pending approval")
+			mon.logCSRsPendingApproval(firstCSRSignerName)
+		}
+
+		if mon.status.KubeletIsRunningOnNode &&
+			!mon.status.SecondCSRSeen &&
+			mon.clusterHasSecondCSRPending() {
+			mon.status.SecondCSRSeen = true
+			mon.logStatus("Second CSR Pending approval")
+			mon.logCSRsPendingApproval(secondCSRSignerName)
+		}
+
+		hasJoined, isReady, err := mon.nodeHasJoinedClusterAndIsReady()
+		if err != nil {
+			logrus.Debugf("nodeHasJoinedClusterAndIsReady returned err: %v", err)
+		}
+
+		if !mon.status.NodeJoinedCluster && hasJoined {
+			mon.status.NodeJoinedCluster = true
+			mon.logStatus("Node joined cluster")
+		}
+
+		if !mon.status.NodeIsReady && isReady {
+			mon.status.NodeIsReady = true
+			mon.logStatus("Node is Ready")
+			// TODO: There appears to be a bug where the node becomes Ready
+			// before second CSR is approved. Log Pending CSRs for now, so users
+			// are aware there are still some waiting their approval even
+			// though the node status is Ready.
+			mon.logCSRsPendingApproval(secondCSRSignerName)
+			cancel()
+		}
+
+		if mon.cluster.API.Rest.IsRestAPILive() {
+			_, err = cluster.MonitorStatusFromAssistedService()
+			if err != nil {
+				logrus.Warnf("Node %s: %s", nodeIPAddress, err)
+			}
+		}
+	}, 5*time.Second, waitContext.Done())
```

Comment:
Not a blocking point for this PR, but a couple of notes that could be addressed separetely in future:

* I'd find more intuitive to have the `cluster.MonitorStatusFromAssistedService` directly within the `cluster.API.Rest.IsRestAPILive()` block.
* Without the burden of using a complete fsm, it could be worth reversing the order of the various states checks and return immediately if satisfied. It could probably also helps in simplifying a little bit the code